src/lib/errors.rs

```diff
@@ -1,28 +1,73 @@
-use std::io::{stderr, Write};
+use std::fmt;
 
 use colored::*;
 use lazy_static::lazy_static;
 
-pub type UfofmtError = Box<dyn std::error::Error + Send + Sync + 'static>;
-pub type Result<T> = std::result::Result<T, UfofmtError>;
+pub(crate) type Result<T> = std::result::Result<T, UfofmtError>;
 
 lazy_static! {
     pub static ref ERROR_INDICATOR: ColoredString = "[ERROR]".red().bold();
     pub static ref OK_INDICATOR: ColoredString = "[OK]".green().bold();
 }
 
-pub(crate) fn print_error(err: &UfofmtError) {
-    let _ = writeln!(stderr(), "{} {}", *ERROR_INDICATOR, err);
+// ufofmt custom error type
+#[derive(Debug)]
+pub(crate) struct UfofmtError {
+    pub(crate) message: String,
+    pub(crate) kind: UfofmtErrorKind,
+}
+
+impl fmt::Display for UfofmtError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> std::result::Result<(), fmt::Error> {
+        match &self.kind {
+            UfofmtErrorKind::Read => {
+                write!(f, "read error: {}", &self.message)
+            }
+            UfofmtErrorKind::Write => {
+                write!(f, "write error: {}", &self.message)
+            }
+            UfofmtErrorKind::InvalidPath => {
+                write!(f, "invalid path error: {}", &self.message)
+            }
+        }
+    }
+}
+
+impl std::error::Error for UfofmtError {}
+
+impl UfofmtError {
+    // Associated functions
+    pub(crate) fn new(kind: UfofmtErrorKind, message: &str) -> Self {
+        Self { kind, message: message.to_owned() }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum UfofmtErrorKind {
```

Comment:
the `ErrorKind` pattern isn't always necessary, and might not be here? I find it's often easier to just have a top-level `Errror` enum, which here could be like,

```rust
enum Error {
    InvalidPath(PathBuf),
    Load(norad::Error),
    Write(norad::Error),
}
```

It's also a nice idiom to keep the inner error intact, where that makes sense. This lets you implement [`std::Error::source`](http://doc.rust-lang.org/1.53.0/std/error/trait.Error.html#method.source) which isn't used much today, but will [hopefully become more important](https://blog.rust-lang.org/inside-rust/2021/07/01/What-the-error-handling-project-group-is-working-towards.html) in the future.

Reply:
👍 

2ac972e1e84381911534c20a873d3edcbb2972e8

I'm trying to keep the ufo path in the error message so added a file path along with the norad error type but otherwise used this approach. Ty!